lib/rosette/integrations/smartling-integration/translation_memory.rb

```diff
@@ -1,43 +1,237 @@
 # encoding: UTF-8
 
+require 'twitter_cldr'
 require 'digest/sha1'
+require 'thread'
 
 module Rosette
   module Integrations
     class SmartlingIntegration < Integration
-      class TranslationMemory
 
+      class TranslationMemory
         DEFAULT_HASH = {}.freeze
+        PLURAL_REGEX = /\.?(zero|one|two|few|many|other)\z/
+        ALLOW_FUZZY = true
+
+        attr_reader :translation_hash, :rosette_config, :repo_config
 
-        def initialize(translation_hash)
+        def initialize(translation_hash, rosette_config, repo_config)
           @translation_hash = translation_hash
+          @rosette_config = rosette_config
+          @repo_config = repo_config
+          @checksum_mutex = Mutex.new
+          @cache_mutex = Mutex.new
         end
 
-        def translation_for(locale, meta_key)
-          all_for_locale = translations_for(locale)
+        def translation_for(locale, phrase)
+          fetch(phrase) do
+            if is_potential_plural?(phrase.meta_key)
+              units = find_plural_translation_for(locale, phrase.meta_key)
+            end
+
+            if !units || units.empty?
+              units = all_translations_for(locale).fetch(phrase.meta_key, [])
+            end
 
-          if translation = all_for_locale[meta_key]
-            translation.key
+            resolve(units, locale, phrase)
           end
         end
 
         def checksum_for(locale)
-          checksums.fetch(locale.code) do
-            digest = Digest::SHA1.new
-            translations = translations_for(locale)
+          @checksum_mutex.synchronize do
+            checksums.fetch(locale.code) do
+              digest = Digest::SHA1.new
+              digest_locale(locale, digest)
+              checksums[locale.code] = digest.hexdigest
+            end
+          end
+        end
 
-            translations.keys.sort.each do |meta_key|
-              digest << meta_key
-              digest << translations[meta_key].key
+        private
+
+        def fetch(phrase)
+          key = (phrase.meta_key || '') + (phrase.key || '')
+          translation_cache.fetch(key) do
+            @cache_mutex.synchronize do
```

Comment:
Unfortunately rosette uses memcached instead of redis, but even if it didn't, there's no way for rosette-integration-smartling to know which kind of cache is being used. We could write our own cache-agnostic mutex that might look something like [this](https://gist.github.com/torkale/2762063). Thoughts?